src/sync/mutex.rs

```diff
@@ -94,15 +94,79 @@ impl<T: ?Sized> Mutex<T> {
                 mutex: self,
             })),
             Err(TryLockError::WouldBlock) => panic!("mutex state out of sync"),
-        }
+        };
+
+        // We need to let other threads in here so they may fail a `try_lock`
+        thread::switch();
+
+        result
     }
 
     /// Attempts to acquire this lock.
     ///
     /// If the lock could not be acquired at this time, then Err is returned. This function does not
     /// block.
     pub fn try_lock(&self) -> TryLockResult<MutexGuard<T>> {
-        unimplemented!()
+        let me = ExecutionState::me();
+
+        let mut state = self.state.borrow_mut();
+        trace!(holder=?state.holder, waiters=?state.waiters, "trying to acquire mutex {:p}", self);
+
+        if let Some(holder) = state.holder {
+            if holder == me {
+                panic!("deadlock! task {:?} tried to acquire a Mutex it already holds", me);
+            }
+        }
+
+        // We don't need a context switch here. There are two cases to analyze.
+        // * Consider that we manage to acquire the lock, but that there is another thread that
+        //   also wants to acquire. At the last context switch (where we were chosen), the other
+        //   thread must have been already runnable and could have been chosen by the scheduler.
+        //   Then that other thread's acquire has a context switch that allows us to run into the
+        //   `WouldBlock` case.
+        // * Consider that we run into the `WouldBlock` case, but that there is another thread
+        //   that wants to release. At the last context switch (where we were chosen), the other
+        //   thread must have been already runnable and could have been chosen by the scheduler.
+        //   Then that other thread's release has a context switch that allows us to acquire the
+        //   lock.
+
+        let result = if state.holder.is_none() {
```

Comment:
Nit: could you rewrite this using a `match` and in the `trace!` on line 158, name the thread that's currently holding the lock?

Reply:
Can we refactor this code path into a common `post_acquire` function? It looks very similar to the `lock` code; refactoring might make any differences more explicit.

Reply:
With the change in vector clock updates, I think a refactoring does not really help.